// tiny-parse-argv/dist/utils.js
/* MAIN */
const castArray = (value) => {
    return Array.isArray(value) ? value : [value];
};
const isBoolean = (value) => {
    return value === true || value === false;
};
const isNil = (value) => {
    return value === null || value === undefined;
};
const isOverridable = (value) => {
    return isNil(value) || isBoolean(value) || value === '';
};
const setNormal = (target, key, value) => {
    if (Array.isArray(target[key])) {
        target[key].push(value);
    }
    else if (isOverridable(target[key])) {
        target[key] = value;
    }
    else {
        target[key] = [target[key], value];
    }
};
const setVariadic = (target, key, value) => {
    const values = castArray(value);
    if (Array.isArray(target[key])) {
        target[key].push(...values);
    }
    else if (isOverridable(target[key])) {
        target[key] = values;
    }
    else {
        target[key] = [target[key], ...values];
    }
};
const uniq = (values) => {
    return Array.from(new Set(values));
};
const without = (values, value) => {
    return values.filter(other => other !== value);
};
const zip = (keys, value) => {
    return Object.fromEntries(Array.from(keys).map(key => [key, value]));
};
/* EXPORT */
export { castArray, isBoolean, isNil, isOverridable, setNormal, setVariadic, uniq, without, zip };
