// tiny-bin/dist/objects/collection.js
/* IMPORT */
import Addon from './addon.js';
import { getClosest } from '../utils.js';
/* MAIN */
class Collection extends Addon {
    constructor() {
        /* VARIABLES */
        super(...arguments);
        this.collection = new Map();
    }
    /* API */
    get(id) {
        return this.collection.get(id);
    }
    getAll() {
        return Array.from(this.collection.values());
    }
    getIds() {
        return Array.from(this.collection.keys());
    }
    getClosest(id) {
        return getClosest(this.getIds(), id);
    }
    getLength() {
        return this.collection.size;
    }
    getOrFail(id) {
        const value = this.get(id);
        if (value)
            return value;
        const closest = this.getClosest(id);
        this.bin.fail(`Not found "${id}"${closest ? `. Did you mean "${closest}"?` : ''}`);
    }
    has(id) {
        return this.collection.has(id);
    }
    register(value) {
        if (this.has(value.id)) {
            this.bin.fail(`Duplicate "${value.id}"`);
        }
        else {
            this.collection.set(value.id, value);
        }
    }
}
/* EXPORT */
export default Collection;
