// @prettier/cli/dist/known.js
//FIXME: This shouldn't be a singleton, but passing it through the whole CLI is a bit of a pain right now
class Known {
    constructor() {
        this.filesPaths = new Set();
        this.filesNames = new Set();
        this.addFilesPaths = (filesPaths) => {
            if (!this.filesPaths.size) {
                this.filesPaths = new Set(filesPaths);
            }
            else {
                for (const filePath of filesPaths) {
                    this.filesPaths.add(filePath);
                }
            }
        };
        this.addFilesNames = (filesNames) => {
            if (!this.filesNames.size) {
                this.filesNames = new Set(filesNames);
            }
            else {
                for (const fileName of filesNames) {
                    this.filesNames.add(fileName);
                }
            }
        };
        this.hasFilePath = (filePath) => {
            return this.filesPaths.has(filePath);
        };
        this.hasFileName = (fileName) => {
            return this.filesNames.has(fileName);
        };
        this.reset = () => {
            this.filesPaths = new Set();
            this.filesNames = new Set();
        };
    }
}
export default new Known();
